Modules/MFT/src/runDigitsRootFileReaderMFT.cxx

```diff
@@ -0,0 +1,144 @@
+#include <Framework/CallbackService.h>
+#include <Framework/ControlService.h>
+#include <Framework/runDataProcessing.h>
+#include <Framework/Task.h>
+#include "ITSMFTBase/Digit.h"
+#include "DataFormatsITSMFT/ROFRecord.h"
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::itsmft;
+
+
+class DigitsRootFileReaderMFT : public o2::framework::Task
+{
+public:
+  //_________________________________________________________________________________________________
+  void init(framework::InitContext& ic)
+  {
+    LOG(INFO) << " In runDigitsRootFileReaderMFT::init ... entering ";
+
+    // open the input file
+    auto filename = ic.options().get<std::string>("mft-digit-infile");
+    mFile = std::make_unique<TFile>(filename.c_str(), "OLD");
+    if (!mFile->IsOpen()) {
+      LOG(ERROR) << "runDigitsRootFileReaderMFT::init. Cannot open the file: " << filename.c_str();
+      ic.services().get<ControlService>().readyToQuit(QuitRequest::All);
+      return;
+    }
+
+  }
+
+  //_________________________________________________________________________________________________
+
+  void run(framework::ProcessingContext& pc)
+  {
+    // get vector of ROF
+    std::unique_ptr<TTree> treeROF((TTree*)mFile->Get("MFTDigitROF"));
+    std::vector<o2::itsmft::ROFRecord> rofs, *profs = &rofs;
+    treeROF->SetBranchAddress("MFTDigitROF", &profs);
+    treeROF->GetEntry(0);
+   
+    // Check if there is a new ROF
+    auto nROFs = rofs.size();
+    if (currentROF >= nROFs) {
+      ///if (currentROF >= 50) {    
+      LOG(INFO) << " runDigitsRootFileReaderMFT::run. End of file reached";
+      pc.services().get<ControlService>().readyToQuit(QuitRequest::All);
+      return;
+    }
+    // prepare the rof output
+    std::vector<o2::itsmft::ROFRecord> oneROFvec;
+    std::copy(rofs.begin()+currentROF, rofs.begin()+currentROF+1, std::back_inserter(oneROFvec));
+    // get the current ROF
+    auto& rof = rofs[currentROF];
+    currentROF++;
+      
+    // get the digits in current ROF
+    // --> get digit branch
+    std::unique_ptr<TTree> treeDig((TTree*)mFile->Get("o2sim"));
+    std::vector<o2::itsmft::Digit> digits, *pdigits = &digits;
+    treeDig->SetBranchAddress("MFTDigit", &pdigits);
+    treeDig->GetEntry(0);
+    // --> find the ranges
+    int index = rof.getROFEntry().getIndex(); // first digit position
+    int nDigitsInROF = rof.getNROFEntries(); // number of digits
+    int lastIndex = index+nDigitsInROF;
+     // --> fill in the corresponding digits
+    std::vector<o2::itsmft::Digit> DigitsInROF;
+    std::copy(digits.begin()+index, digits.begin()+lastIndex, std::back_inserter(DigitsInROF));
+
+    // fill in the message
+    //   LOG(INFO) << " runDigitsRootFileReaderMFT::run. In this ROF there are  " << DigitsInROF.size() << " digits";
+    pc.outputs().snapshot(Output{"MFT", "DIGITS", 0, Lifetime::Timeframe}, DigitsInROF);
+    pc.outputs().snapshot(Output{"MFT", "MFTDigitROF", 0, Lifetime::Timeframe}, oneROFvec);
```

Comment:
Since these vectors are destructed anyway when going outside of the scope of this functions, consider using `adopt` instead of `snapshot`. The first one takes a pointer to your object, takes ownership and deletes it on its own (so you have to allocate it dynamically) - no unnecessary copies involved. The latter copies the object and sends that further.
You might want to try it out, if you care about performance here.

Reply:
Thanks, I implemented the suggestion.

Reply:
When I try to change snapshot for adopt I get an error when compiling QC. Here is the log: https://cernbox.cern.ch/index.php/s/rv2rKED9R6rqtjX

Would you know how to fix it?

Reply:
`adopt` needs a pointer, not a value. So first, you have to allocate your vector using `new`, so:
```
std::vector<o2::itsmft::Digit>* DigitsInROF = new std::vector<o2::itsmft::Digit>();
```
instead of
```
std::vector<o2::itsmft::Digit> DigitsInROF;
```
And then you have to pass that pointer to `adopt`.
If that doesn't compile, then it is a bug in `DataAllocator` and we can complain to its authors.

Reply:
I have allocated the vector as you suggested and also I am using *DigitsInROF to refer to the vector. But I still seem to miss something as I get an error when using adopt. 

Here is the log file: https://cernbox.cern.ch/index.php/s/anJgUh6S09BMQM8
Here is the modified macro:https://cernbox.cern.ch/index.php/s/PX2Gax5m76ZHavu

Reply:
The code you have sent me has still `snapshot`, not `adopt`.
But by looking at the logs, the problem is that you are still passing a *value*, not a *pointer*. You do:
```
pc.outputs().adopt(Output{ "MFT", "DIGITS", 0, Lifetime::Timeframe }, *DigitsInROF);
```
As `DigitsInROF` is a pointer now. if you write `*DigitsInROF`, you get the value under that pointer.
Try removing the asterisk, then you will pass the pointer.

Reply:
Sorry for the mess up with the file. I tried what you suggested but it also fails. 

Here is the, hopefully correct, file: https://cernbox.cern.ch/index.php/s/PX2Gax5m76ZHavu
Here is the new log: https://cernbox.cern.ch/index.php/s/kYb57xPDtoScqyv

Reply:
@matthiasrichter is `adopt` expected to work here? we are trying to do:
```
std::vector<o2::itsmft::ROFRecord>* oneROFvec = new std::vector<o2::itsmft::ROFRecord>();
std::vector<o2::itsmft::Digit>* DigitsInROF = new std::vector<o2::itsmft::Digit>();
...
pc.outputs().adopt(Output{ "MFT", "DIGITS", 0, Lifetime::Timeframe }, DigitsInROF);
pc.outputs().adopt(Output{ "MFT", "MFTDigitROF", 0, Lifetime::Timeframe }, oneROFvec);
```
but the compilation fails as in the logs from the comment above.

Reply:
Ok, since it is hard to get anyone's response recently, let's leave it as it is for now. We can improve it when the problem with `adopt` is understood.

Reply:
Ok, thank you for all the help.